python/ray/dag/compiled_dag_node.py

```diff
@@ -269,27 +282,66 @@ def _preprocess(self) -> None:
             FunctionNode,
             InputAttributeNode,
             InputNode,
+            MultiInputNode,
             MultiOutputNode,
         )
 
-        self.input_task_idx, self.output_task_idx = None, None
+        self.input_task_idxs, self.output_task_idx = [], None
         self.actor_task_count.clear()
 
+        # If InputAttributeNodes are used, remove the InputNode and
+        # InputAttributeNodes and pack their info into a MultiInputNode,
+        # then reorder nodes
+        has_attr_node = any(
+            [
+                isinstance(task.dag_node, InputAttributeNode)
+                for _, task in self.idx_to_task.items()
+            ]
+        )
+        if has_attr_node:
+            self.counter = 0
+            tasks = [task for _, task in self.idx_to_task.items()]
+            self.idx_to_task.clear()
+            self.dag_node_to_idx.clear()
+            arg_list = []
+            arg_idx = 0
+            # Add attribute nodes first
+            for task in tasks:
+                if isinstance(task.dag_node, InputAttributeNode):
+                    arg_list.append(task.dag_node._key)
+                    if isinstance(task.dag_node._key, str):
+                        self.kwarg_to_idx[task.dag_node._key] = arg_idx
+                    arg_idx += 1
+            multi_input_node = MultiInputNode(*arg_list)
+            self._add_node(multi_input_node)
+            for task in tasks:
+                if isinstance(task.dag_node, InputAttributeNode):
+                    for down_stream_idx in task.downstream_node_idxs:
+                        tasks[down_stream_idx].args = (multi_input_node,)
+                elif not isinstance(task.dag_node, InputNode):
+                    self._add_node(task.dag_node)
+            for idx, task in self.idx_to_task.items():
+                task.dag_node.set_args(
+                    tuple(
+                        [
+                            multi_input_node
```

Comment:
I didn't really understand this part -- don't we need to keep track of which arg the task is supposed to read from the multi-args?

Reply:
we track that by arg read index